src/Messages/OWSMessageSender.m

```diff
@@ -212,18 +210,17 @@ - (instancetype)initWithMessage:(TSOutgoingMessage *)message
             return;
         }
 
-        // Ensure we call the success or failure handler exactly once.
-        @synchronized(strongSelf)
-        {
-            OWSCAssert(!strongSelf.hasCompleted);
-            strongSelf.hasCompleted = YES;
-        }
-
         [strongSelf.message updateWithSendingError:error];
 
         DDLogDebug(@"%@ failed with error: %@", strongSelf.tag, error);
         aFailureHandler(error);
-        [strongSelf markAsComplete];
+
+        // Ensure we call the success or failure handler exactly once.
+        @synchronized(strongSelf)
```

Comment:
nit: I think we could just do the check *in* `markAsComplete` and dedupe the two asserts blocks.